OpenTrons/opentrons-api
# feat(protocol-designer, step-generation, shared-data): 96-channel tip adapter + well selection support

closes RAUT-678 RAUT-679 RAUT-697 RAUT-755

this pr is now unblocked ✨ 

# Overview

This adds support for 2 main things for PD 96-channel:
1. 96-channel tiprack adapter support
2. 96-channel well selection support

Sorry this ended up doing a lot in 1 PR! 

# Test Plan

For tiprack adapter:

1. turn on the 96-channel dev feature flag
2. create a Flex protocol and add the 96-channel
4. go to deck setup. Observe that the tiprack on the deck has an adapter underneath it (the adapter is just a white svg rectangle thing underneath the tiprack)
5. try duplicating the tiprack, the duplicate should NOT have an adapter underneath
6. select an empty deck slot and add another tiprack, it should NOT render the adapter underneath
7. select an empty deck slot and add the 96-channel adapter (under the adapter section), it should render only the adapter. Check that the only compatible labware is the tiprack.
8. add another adapter with no tiprack, try to drag and drop other labwares onto it, make sure only the tiprack is supported to drag and drop onto the adapter but that the tiprack also can be directly on the deck.
9. add a move labware command, make sure only the tiprack can move on top of the adapter .

For 96-channel well selection:

1. create a Flex protocol with the 96-channel and a tiprack and add the trash bin (the waste chute isn't fully wired up for dispensing/dropping tip into so a trash bin is required for now). Go the deck map and add a reservoir, a 96 well plate, an irregular plate, and the 384 well plate.
2. try to mix/transfer from out of each plate. See that they all work except for the irregular plate. See that the correct amount of wells are highlighted (all wells except for the 384 well plate which should highlight every other well for up to 96 wells.)
4. now create a new protocol with a single and 8-channel pipette. See that transferring/mixing works as usual and highlights either 1 or 8 wells.

# Changelog

- `createContainer` thunk to include an optional `adapterUnderLabwareURI` or `adapterId` in order to allow for duplicating the labware with the adapter underneath it. 
- extend the deck setup components to have knowledge of if the 96-channel is in the protocol
- extend `getWellNamePerMultiTip` and affecting utils to have knowledge of if the pipette has 96-channels 
- add a few helper functions to shared-data (moved `orderedWells` out of step-generation and into shared-data and created a `get96Channel384WellPlateWells` util that grabs the arrays of wells the 96-channel can aspirate from in a 384 well plate)
- fix tests
- create a `missingAdapter()` error creator to be plugged in for a follow up PR

# Review requests

see test plan

# Risk assessment

low-ish, this changes a bunch of logic
